### ISSUE TITLE
[C# / Feature Request] Adding base_namespace option to grpc_csharp_plugin

### ISSUE THREAD

**Issue by Falco20019** [grpc/grpc#13724]
### What version of gRPC and what language are you using?
Grpc.Tools v1.7.3 from the C# NuGet
 
### What operating system (Linux, Windows, …) and version?
Windows 10 64-bit
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
Microsoft Visual Studio Professional 2017 v15.4.3
 
### What did you do?
Tried providing protoc with `--grpc_out=%output_path% --grpc_opt=base_namespace` and `--grpc_out=base_namespace:%output_path%`.
 
### What did you expect to see?
Get the generated service files sorted into the same folder structure analogous to `--csharp_opt=base_namespace`.
 
### What did you see instead?
The option is not available. Without the option, every service file is put into the base output folder without any structure, leading to conflicts in more complex projects with a lot of services. Also leading to a not very nice overview.

**Comment by ronnylt** [comment]
Is there any workaround for this?  As commented, all services are generated in the base output folder.

**Comment by Falco20019** [comment]
Sadly none that I know off :( we still have to make sure ourselves that no one in any of our teams uses the same name

**Comment by ronnylt** [comment]
I wonder how they generate googleapis, see for example https://github.com/googleapis/google-cloud-dotnet/tree/master/apis/Google.Cloud.Bigtable.V2

**Comment by Falco20019** [comment]
@ronnylt This should be generated using [GAPIC](https://github.com/googleapis/gapic-generator/blob/master/README.md). Also based on protoc but using a separate generator plugin.

**Comment by kkm000** [comment]
@Falco20019, I am looking at a possibility of fixing this and #17672 in the release after next. Just to confirm, you would be fine if protoc put the files *still under the `obj/` directory*, but arranged them according to namespaces with `--csharp_opt=base_namespace`--am I understanding your suggestion correctly?

**Comment by Falco20019** [comment]
@kkm000 Yes you are right. We just have to avoid naming conflicts for cases like `protos/dir1/bar.proto` and `protos/dir2/bar.proto` being generated both as `obj/Bar.cs` and `obj/BarGrpc.cs` instead of `obj/dir1/Bar[Grpc].cs` and `obj/dir2/Bar[Grpc].cs`.

Right now generation would fail completely with the messages:
```
MSB3105: The item "obj\Debug\netstandard2.0\Bar.cs" was specified more than once in the "Sources" parameter.  Duplicate items are not supported by the "Sources" parameter.
MSB3105: The item "obj\Debug\netstandard2.0\BarGrpc.cs" was specified more than once in the "Sources" parameter.  Duplicate items are not supported by the "Sources" parameter.
```

**Comment by kkm000** [comment]
I see. I'll look how possible it is to match the protoc's logic of naming paths. Might be tricky.

As a workaround, if your grab your proto files recursively, you may leverage the [`%(RecursiveDir)` metadatum](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-well-known-item-metadata) to automatically add each file's only automatically recursed portion of path to its output location, with a default for the whole project, with e. g.
```xml
  <Protobuf Include="../../protos/**/*.proto" />
  <ItemDefinitionGroup>
    <Protobuf>
      <OutputDir Condition="'%(Protobuf.OutputDir)' == '' ">$(Protobuf_OutputPath)%(RecursiveDir)</OutputDir>
    </Protobuf>
  </ItemDefinitionGroup>
```
In this setup, e. g.  `../../protos/dir1/bar.proto` will have its RecursiveDir set to `dir1/`, i.e. the portion of path matching the `**` in the Include expression. (there is also Exclude pattern if you want to skip some files and/or directories).

Depending on how many conflicts you have, there are other options, like adding files in groups, e. g.
```xml
  <Protobuf Include="protos/dir1/*.proto;protos/dir2/*.proto" OutputPath="$(Protobuf_OutputPath)out1" />
  <Protobuf Include="protos/dir3/**/*.proto;protos/dir4/*.proto" OutputPath="$(Protobuf_OutputPath)out2" />
```
There should be a workaround, more or less tricky.

`$(Protobuf_OutputPath)` is set to the project's obj/ directory by default.

**Comment by Falco20019** [comment]
Not really an easy option for us since we provide it as part of a framework and have lots of csproj files. The workaround would be too complicated for a lot of our devs.

Protoc is using the package path to generate the path for it. I will give it a look next week, maybe I can help with the PR.

**Comment by kkm000** [comment]
Thank you, I'll check what we can do, too. A significant problem here is we have to replicate (and then support in sync) the protoc's logic exactly in the tooling, and expect the right files with correct names in exact locations where protoc drops them; it we miss the target, the project may rebuild every time you hit build. protoc with `--csharp_opt=base_namespace` derives subdirectories from file *contents* (since the namespace is declared inside a file), where we cannot even peek into without parsing the file. So this is quite a tricky issue.

**Comment by Falco20019** [comment]
I think this response partly belongs to #17672.

The grpc_csharp_plugin should already get the parsed content by protoc if I remember correctly.

Belonging to Grpc.Tools: We could also create a output folder under /obj (like /obj/protos) and include all generated files from there. This would reduce the need to look the contents up. I will give it a look on Monday or Tuesday to get a feeling for the problem since I‘m definetly not so deep into MSBuild yet than you are and might just miss some of problems involved.

**Comment by Falco20019** [comment]
Not stale, still needed.

**Comment by EraYaN** [comment]
This should probably also match the naming with all the transformations `two-words` -> `TwoWords` for the file paths. Now only underscores seem to transform correctly. `two_words` -> `TwoWords`.

**Comment by Falco20019** [comment]
I implemented it now by using the same logic as was used for the `csharp_generator` in `protoc`. I tested it locally for all paths and it seems to work like a charm. I even added the returned error message as the current version only returns `false` without setting the `error` variable.

As second step, I prepare another PR for the Grpc.Tools adjustment.

**Comment by Falco20019** [comment]
Will be fixed for me using #17672
